Pulse-level control instructions (#184)
Qiskit/openqasm: source/language/openpulse.rst

OpenPulse Grammar
=================

OpenQASM allows users to provide the target system's implementation of quantum operations
with ``defcal`` blocks . Calibration grammars are open to extension for system implementors. In

> this document, we outline one such grammar, OpenPulse, which maybe selected within a supporting

Reviewer: ```suggestion
this document, we outline one such grammar, OpenPulse, which may be selected within a supporting
```

compiler through the declaration defcalgrammar "openpulse" <VERSION>.

This grammar is motivated by the original OpenPulse specification, a JSON wire-format for
pulse-level quantum programs defined in the paper `Qiskit Backend Specifications for OpenQASM
and OpenPulse Experiment`_:cite:`mckay2018`. The textual format described
here has several advantages over the original JSON format:

> - Improved readability

Reviewer: Extremely minor but should be consistent on periods vs no periods at the end of the bullets

Reviewer 2: Fixed in upcoming 597c3e5.

- Pulse timing is based on instruction ordering and works with programs containing branching
  control flow
- Reusable gate calibrations enable more succinct calibration descriptions.
- Pulse definitions are declared as a calibration for individual circuit instructions attached to
  physical qubits enabling the microcoding of gate level operations.
- Richer ability to compose complex pulses through natural DSP-like operations.
- Clearly defined relationship between pulses, channels and the phases of compilation to
  hardware resources.
- Use of multiple frames on a single channel at the same time (see :ref:`Geometric gate`)

Openpulse provides a flexible programming model that should extend to many quantum control schemes

> and hardware. At the core of the OpenPulse grammar are the concepts of ``frames``.

Reviewer: ```suggestion
and hardware. At the core of the OpenPulse grammar are the concepts of ``frames``, ``waveforms`` and ``channels``.
```

Reviewer 2: Addressed in upcoming commit e374390.

``frames`` are used to schedule the playing of a ``waveform`` or capturing of data

> via a ``channel`` on the target device. A channel is a software abstraction which allows the

Reviewer: ```suggestion
via a ``channel`` on the target device. A channel is a software abstraction which allows the
```
Is the channel the abstraction or is it the `frame`? I thought the channels were supposed to map pretty closely to HW resources (as stated in the next section). They could still be abstractions, but I thought they were supposed to mirror input and output ports relatively closely.

> via a ``channel`` on the target device. A channel is a software abstraction which allows the

Reviewer: ```suggestion
via a ``channel`` on the target device. A ``channel`` is a software abstraction which allows the
```

Reviewer 2: Addressed in upcoming commit 28d7aa4.

programmer to be agnostic to complexities of the device's underlying pulse generation hardware. It

> is the responsibility of the target device's compiler to map ``frames`` to the applied channels in

Reviewer: ```suggestion
is the responsibility of the target device's compiler to map ``frames`` to the applied ``channel``'s in
```

Reviewer 2: Fixed in af62a50.

interactions with that qubit. A channel may also have many qubits. For instance,
a channel could manipulate the coupling between two neighboring qubits, or
could even reference multiple qubits in a chain.

Channels are vendor-specific and device-specific. It is expected that vendors
of quantum hardware provide the appropriate channel names and qubit mappings
as configuration information to end users. Each channel may also have associated
static settings, such as local-oscillator frequencies, which do not vary
throughout program execution. Again it is expected that vendors of quantum
hardware provide a method for manipulating those static settings if appropriate.

There are two kinds of channels: transmit channels (sending input to a quantum
device) and receive channels (reading output from a quantum device).

A channel is only used to specify the physical resource on which to play a pulse or from which
to capture data. This specification should be done by providing a mapping between a qubit list +
name and the configured hardware channel. The hardware can then be accessed as
OpenPulse ``txchannel``/``rxchannel``'s via "get" functions.

.. code-block:: c

>     txch(qubit q0, ..., qubit qn, str name) -> txchannel  // get transmit channel

Reviewer: Should we add `$` (ie `qubit $q0`) to be explicit here about the physical qubits?

Reviewer 3: We should also change these to `channel` as was decided offline (and as has already been done in some of the examples).

Reviewer 2: Fixed in e5b0181.

    rxch(qubit q0, ..., qubit qn, str name) -> rxchannel  // get receive channel

The qubits must be **physical** qubits. Furthermore, ordering of qubits is important. For instance,
``txch($0, $1, "control")`` and ``txch($1, $0, "control")`` may be used to implement distinct
cross-resonance gates. It is also possible to access a channel by its full name, without supplying

> any qubits, if that has between implemented by the vendor. For instance, ``txch("<channel_name>")``

Author: Further typo: it should be "if that has _been_ implemented" 😄

> any qubits, if that has between implemented by the vendor. For instance, ``txch("<channel_name>")``

Reviewer 2: Fixed in e19a37f.

may refer to a transmit channel with an arbitrary name.

.. code-block:: c

    txchannel d0 = txch($0, "drive");  // channel for driving at qubit $0's freq
    txchannel cr1_2 = txch($1, $2, "coupler");  // channel for a coupler between two qubits

>     txchannel m2 = txch($2, "measure");  // channel for transmitting measurement stimulus

Reviewer: Add name-only example of channel declaration...

Reviewer 2: Fixed in da29859.

  frame (i.e. with the carrier supplied by the ``frame``). Another motivation is to more naturally
  implement a "virtual Z-gate", which does not require a physical pulse but rather shifts the phase
  of all future pulses on that frame.

The frame is composed of three parts:

1. A frequency ``frequency`` of type ``float``.
2. A phase ``phase`` of type ``angle``.
3. A time which is manipulated implicitly and cannot be modified other
   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
   is determined by the channel on which the frame is played (see :ref:`Timing` section).

The ``frame`` type is a virtual resource and the exact precision of these parameters is
hardware specific. It is thus up to the compiler to choose how to implement the required
transformations to physical resources in hardware (e.g. mapping multiple frames to a
single NCO).

Frame Construction
~~~~~~~~~~~~~~~~~~

Frames can be constructed using the ``newframe`` command e.g.

.. code-block:: javascript

  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)

> would construct a frame with frequency `f`  5 GHz and phase :math:`\theta` 0.0. When
> instantiated, the frame time starts at 0. ``frame``s can also be copied using the
> ``copyframe`` command

Reviewer 3: ```suggestion
would construct a frame with a frequency of 5 GHz and a phase of 0.0. When
instantiated, the frame time starts at 0. ``frame``\s can also be copied using the
``copyframe`` command
```

Reviewer 2: Fixed in bc580c6.

.. code-block:: javascript

  frame driveframe1 = newframe(5e9, 0.0);
  frame driveframe2 = copyframe(driveframe1);
  driveframe2.phase = driveframe1.phase + pi/2;

> This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` with the same

Reviewer: ```suggestion
This will generate a :math:`\pi/2` phase incremented copy of ``driveframe1`` with the same
```

Reviewer 2: Fixed in fa41cff.

time as `driveframe1`.

Frame manipulation
~~~~~~~~~~~~~~~~~~

The (frequency, phase) tuple of a frame can be manipulated throughout program
by referencing ``.frequency``, and ``.phase``. Operations must be
appropriate for the respective type, ``float`` for frequency and ``angle`` for
phase. Again, the exact precision of these calculations is hardware specific.

Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
``driveframe``:

.. code-block:: javascript

   // Example 1: Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
   cal {
     driveframe.phase += pi/4;
   }

   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
   cal {

>      frame[8] rz_frames;

Reviewer: Are we allowing frame registers? The array type is not yet defined in openqasm, but when it is I was under the impression it was going to be something like `array[frame, 8]` (due to the fact that many classical types contain both array length and precision).

Author: Good point, we need to define arrays which is what this is.

Reviewer 2: Updated array frames here e858e5f. Left a comment noting we're stilling trying to resolve this.

>      frame[8] rz_frames;

Reviewer 4: To be revisited when we intro `angle` syntax, but I'd prefer that `frame` not take a width, and you instead re-write this line as `angle[frame, 8] rz_frames`.

Author: I think you mean `array[frame, 8] rz_frames`, but yes I agree.

Reviewer 2: Updated array frames here e858e5f. Left a comment noting we're stilling trying to resolve this.

     frame[0] = newframe(...);
     // and so on
   }

>    defcal rz(angle[20] theta) $q {

Reviewer: Are `defcal`'s on qubit registers allowed? I thought the `$q` syntax would refer to an individual physical qubit anyway...

Author: defcals can only be associated with gates which broadcast to single qubits. See https://github.com/Qiskit/openqasm/issues/215

     rz_frames[q].phase -= theta;
   }

Manipulating frames based on the state of other frames is also permitted:

.. code-block:: javascript

   // Swap phases between two frames
   const temp = frame1.phase;
   frame1.phase = frame2.phase;
   frame2.phase = temp;

Waveforms
---------

Waveforms are of type ``waveform`` and can either be:

- An array of complex samples which define the points for the waveform envelope
- An abstract mathematical function representing a waveform. This will later be
  materialized into a list of complex samples, either by the compiler or the hardware
  using the parameters provided to the pulse template.

A value of type ``waveform`` is retrieved by explicitly constructing the complex samples
or by calling one of the built-in waveform template functions. The latter are initialized by

> assigning a ``waveform`` to the result of a (kernel) function call. Note that each of these

Reviewer 3: `kernel` --> `extern`

Reviewer 2: Addressed in e388e06.

kernel functions takes a type ``length`` as a first argument, since waveforms need to have a
definite length. Using the hardware dependent ``dt`` unit is recommended for this length,
since the compiler may need to down-sample a higher precision waveform to physically realize it.

.. code-block:: javascript

   // arbitrary complex samples
   waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];

   // amp is waveform amplitude at center
   // l is the overall length of the waveform
   // sigma is the standard deviation of waveform

>    kernel gaussian(complex[size] amp, length l, length sigma) -> waveform;

Reviewer: Says in the description that `l` is the first parameter? Not a big deal, but should try to be consistent...

Author: From @levbishop comment here: https://github.com/Qiskit/openqasm/issues/121#issuecomment-769210488 we should consider putting the length in square brackets at the beginning.

Reviewer: Agreed.

>    kernel gaussian(complex[size] amp, length l, length sigma) -> waveform;

Reviewer 4: `kernel` here and below should be `extern`.

Reviewer 2: Resolved in e388e06.

   // amp is waveform amplitude at center
   // l is the overall length of the waveform
   // square_width is the width of the square waveform component
   // sigma is the standard deviation of waveform
   kernel gaussian_square(complex[size] amp, length l, length square_width, length sigma) -> waveform;

   // amp is waveform amplitude at center
   // l is the overall length of the waveform
   // sigma is the standard deviation of waveform
   // beta is the Y correction amplitude, see the DRAG paper
   kernel drag(complex[size] amp, length l, length sigma, float[size] beta) -> waveform;

   // amp is waveform amplitude
   // l is the overall length of the waveform
   kernel constant(complex[size] amp, length l) -> waveform;

   // amp is waveform amplitude
   // l is the overall length of the waveform
   // frequency is the frequency of the waveform
   // phase is the phase of the waveform
   kernel sine(complex[size] amp, length  l, float[size] frequency, angle[size] phase) -> waveform;

We can manipulate the ``waveform`` types using the following signal processing functions to produce
new waveforms (this list may be updated as more functionality is required).

> .. code-block:: javascript

Reviewer 3: One useful function I forgot about keeping around was `scale`. I think this is valuable enough and it's even employed in the "Geometric gate" example.

Reviewer 2: Addressed in 36bafc4.

    // Multiply two input waveforms entry by entry to produce a new waveform
    // :math:`wf(t_i) = wf_1(t_i) \times wf_2(t_i)`
    kernel mix(waveform wf1, waveform wf2) -> waveform;

    // Sum two input waveforms entry by entry to produce a new waveform
    // :math:`wf(t_i) = wf_1(t_i) + wf_2(t_i)`
    kernel sum(waveform wf1, waveform wf2) -> waveform;

>     // Add a relative phase to a waveform (ie multiply by :math:`e^{\imag \theta}`

Reviewer: ```suggestion
    // Add a relative phase to a waveform (ie multiply by :math:`e^{\imag \theta}`)
```

    kernel phase_shift(waveform wf, angle ang) -> waveform;

Play instruction
----------------

Waveforms are scheduled using the ``play`` instruction. These instructions may
only appear inside a ``defcal`` block and have three required parameters:

- the channel on which to play the pulse
- a value of type ``waveform`` representing the waveform envelope
- the frame to use for the pulse

.. code-block:: javascript

  play(channel chan, waveform wfm, frame fr)

For example,

.. code-block:: javascript

  defcal play_my_pulses {
   // Play a 3 sample pulse on the tx0 channel
   play(tx0, [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j], driveframe);

   // Play a gaussian pulse on the tx1 channel
   frame f1 = newframe(q1_freq, 0.0);
   play(tx1, gaussian(...), f1);
  }

Capture Instruction
-------------------

Acquisition is scheduled by a ``capture`` instruction. This is a special

> ``kernel`` function which is specified by a hardware vendor. The measurement

Reviewer 4: ```suggestion
``extern`` function which is specified by a hardware vendor. The measurement
```

process is difficult to describe generically due to the wide variety of
hardware and measurement methods. Like the play instruction, these instructions
may only appear inside a ``defcal`` block!

The only required parameters are the ``channel`` and the ``frame``.

The following are possible parameters that might be included:

- A "duration" of type ``length``, if it cannot be inferred from other parameters.
- A "filter" of type ``waveform``, which is dot product-ed with the measured IQ to distill the
  result into a single IQ value

Again it is up to the hardware vendor to determine the parameters and write a

> kernel definition at the top-level, such as:

Reviewer 4: Again, all these `kernel`s need to become `extern`s.

Reviewer 2: Addressed in same commit as others.

.. code-block:: javascript

>    // Minimum requirement

Reviewer 2: Addressed in 2a4382d.

   kernel capture(channel chan, frame output) -> complex[32];

   // A capture command that returns an iq value
   kernel capture(channel chan, waveform filter, frame output) -> complex[32];

   // A capture command that returns a discrimnated bit
   kernel capture(channel chan, waveform filter, frame output) -> bit;

   // A capture command that returns a raw waveform data
   kernel capture(channel chan, length len, frame output) -> waveform;

> 

Reviewer 2: Addressed in 5f4ec57.

The return type of a ``capture`` command varies. It could be a raw trace, ie. a
list of samples taken over a short period of time. It could be some averaged IQ
value. It could be a classified bit. Or it could even have no return value,
pushing the results into some buffer which is then accessed outside the program.

For example, the ``capture`` instruction could return raw waveform data that is then

> discriminated using user-defined boxcar and discrimnation ``kernel``s.

Reviewer: ```suggestion
discriminated using user-defined boxcar and discrimination ``kernel``s.
```

> discriminated using user-defined boxcar and discrimination ``kernel``s.

Reviewer 3: ```suggestion
discriminated using user-defined boxcar and discrimination ``extern``\s.
```

.. code-block:: javascript

    // Use a boxcar function to generate IQ data from raw waveform
    kernel boxcar(waveform input) -> complex[64];
    // Use a linear discriminator to generate bits from IQ data
    kernel discriminate(complex[64] iq) -> bit;

    defcal measure $0 -> bit {
        // Define the channels
        txchannel m0 = txch($0, "measure");
        rxchannel cap0 = rxch($0, "capture");

        // Force time of carrier to 0 for consistent phase for discrimination.
        frame stimulus_frame = newframe(5e9, 0);
        frame capture_frame = newframe(5e9, 0);

        // Apply measurement stimulus
        waveform meas_wf = gaussian_square(1.0, 16000dt, 262dt, 13952dt);

        // Play the stimulus
        play(m0, meas_wf, stimulus_frame);
        // Align measure and capture channels
        barrier(stimulus_frame, capture_frame);
        // Capture transmitted data after interaction with measurement resonator
        // kernel capture(channel chan, frame capture_frame) -> waveform;

>         waveform raw_output = receive(cap0, meas_pulse.duration);

Reviewer: What is `receive`? Believe this is lurking from past drafts. Also do we require the `duration` for the `capture`? How else do we know when to stop collecting data?

Reviewer 3: Yeah this is a stale example. I think it should read more like
```
// extern capture(channel chan, length len, frame output) -> waveform;
waveform raw_output = capture(cap0, meas_pulse.duration, capture_frame)
```

Reviewer 2: Fixed in fa97dfa.

Each frame maintains its own "clock". When a pulse is played the clock for
that frame advances by the length of the pulse.

For frames, everything behaves analogous to qubits in the
`Delays <delays.html>`_ section of this specification. There are however some
small differences.

The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
instruction may also take a list of frames instead of a list of qubits and aligns the time
of the clocks given as arguments.

``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
meaning that those clocks are guaranteed to be aligned at the start of the block.
These blocks also need to have a well-defined length, similar to the ``boxas`` block.

.. code-block:: javascript

   cal {
     waveform p = ...; // some 100dt waveform
     frame driveframe1 = newframe(5.0e9, 0);
     frame driveframe2 = newframe(6.0e9, 0);
   }

   defcal aligned_gates {
     // driveframe1 and driveframe2 used in this defcal, so clocks are aligned
     play(tx0, p, driveframe1);
     delay[20dt] driveframe1;
     // Clocks now unaligned by 120dt, so we use a `barrier` to re-align
     barrier(driveframe1, driveframe2);

>      // `driveframe2` will now play a pulse 120dt after `driveframe1` finishes playing

Reviewer 3: I think this should either read 
```
// `driveframe2` will now play a pulse 20dt after `driveframe1` finishes playing
```
or 
```
// `driveframe2` will begin playing its pulse 120dt after `driveframe1` begins playing
```

Reviewer 2: Fixed in 3004f53.

     play(tx0, p, driveframe2);
   }

Examples
--------

Cross-resonance gate
~~~~~~~~~~~~~~~~~~~~

.. code-block:: javascript

  cal {
     frame frame0 = newframe(5.0e9, 0);
  }

>   defcal cross_resonance $0, $1 {

Reviewer 3: Should we throw the `d0` and `d1` channels into the `cal`?

Reviewer 2: I don't think so as this demos lookup of channels for now. Still not a huge fan of this approach.

      // Access globally defined channels

>       channel d0 = txch($0, "drive");

Reviewer: ```suggestion
      txchannel d0 = txch($0, "drive");
```

Author: Ignoring since the decision was to use channel instead of txchannel/rxchannel

>       channel d1 = txch($1, "drive");

Reviewer: ```suggestion
      txchannel d1 = txch($1, "drive");
```

      waveform wf1 = gaussian_square(1., 1024dt, 128dt, 32dt);
      waveform wf2 = gaussian_square(0.1, 1024dt, 128dt, 32dt);

      // phase update some virtual Z gate

>       frame0.phase += pi/2;

Reviewer: Clarification: This affects `frame0` if applied on any other gate, right? Since it's a global?

Author: Yes that's right

      /*** Do pre-rotation ***/
      {...}

      // generate new frame for second drive that is locally scoped
      frame temp_frame = copyframe(frame0);
      temp_frame.phase = frame0.phase + pi/2;

>       play(d0, wf1, frame0);

Reviewer: Don't we need a third pulse, also on `d0`, but with a new `frame` to play at the CR frequency? (the difference between qubit 0 and 1 resonant frequencies)

Reviewer 2: Updated example in  94bb38e.

      play(d1, wf2, temp_frame);

      /*** Do post-rotation ***/
      {...}
  }

Geometric gate
~~~~~~~~~~~~~~

.. code-block:: javascript

  float[32] fq_01 = 5e9; // hardcode or pull from some function

>   float[32] anharm = 300e6; // hardcode or pull from some function

Reviewer 3: Should we throw these in the `cal`?

Reviewer 2: Added to cal block in 0a1bb4f.

  cal {
      frame frame_01 = newframe(fq_01, 0);
      frame frame_12 = newframe(fq_01 + anharm, 0);
  }

>   defcal geo_gate(angle[32] theta) $0 {

Reviewer: ```suggestion
  defcal geo_gate(angle[32] theta) $q {
```

      // theta: rotation angle (about z-axis) on Bloch sphere

      // Access globally defined channels
      tx_channel dq = txch($q, “drive”);

      // Assume we have calibrated 0->1 pi pulses and 1->2 pi pulse
      // envelopes (no sideband)
      waveform X_01 = {...};
      waveform X_12 = {...};
      float[32] a = sin(theta/2);
      float[32] b = sqrt(1-a**2);

      // Double-tap

>       play(dq, scale(a, X_01), frame_01);

Reviewer: Does `scale` function exist anymore? Don't see it...

Reviewer 3: Yeah, it's gone but I think it's valuable to re-introduce. Made a comment to that effect in the "Waveform" section.

Reviewer 2: Reintroduced.

      play(dq, scale(b, X_12), frame_12);
      play(dq, scale(a, X_01), frame_01);
      play(dq, scale(b, X_12), frame_12);
  }

Neutral atoms
~~~~~~~~~~~~~

In this simple example, the signal chain is composed of two electro-optic modulators (EOM) and
an acousto-optic deflector (AOD). The EOMs put sidebands on the laser light while the AOD diffracts
the light in an amount proportional to the frequency of the RF drive. This example was chosen

> because it is similar in spirit to the work by Levine et al.:cite:`levine2019` except that phase

Reviewer 3: ```suggestion
because it is similar in spirit to the work by Levine et al._:cite:`levine2019` except that phase
```

Reviewer 2: Fixed in 6565480.

control is exerted using virtual Z gates on the AODs -- requiring frame tracking of the qubit
frequency yet application of a tone that maps to the qubit position (i.e. requires the use of a
sideband).

The program aims to perform a Hahn echo sequence on q1, and a Ramsey sequence on q2 and q3.

.. code-block:: javascript

  defcal neutral_atoms {
    // Access globally defined channels

>     channel eom_a_channel = txch(0, "eom_a");

Reviewer: Make these `txchannel`'s.

Reviewer: Also should they be physical qubits? Ie `0`->`$0`, et...

Reviewer 3: I think the agreement offline was to use `channel` . Also, yes they should be physical qubits.

Reviewer 2: 0f5972b

    channel eom_a_channel = txch(1, "eom_b");
    channel aod_channel = txch(0, "aod");

    // Define the Raman frames, which are detuned by an amount Δ from the  5S1/2 to 5P1/2 transition
    // and offset from each other by the qubit_freq
    frame raman_a_frame = newframe(Δ, 0.0)
    frame raman_b_frame = newframe(Δ-qubit_freq, 0.0)

    // Three copies of qubit freq to track phase of each qubit
    frame q1_frame = newframe(qubit_freq, 0)
    frame q2_frame = newframe(qubit_freq, 0)
    frame q3_frame = newframe(qubit_freq, 0)

    // Generic gaussian envelope
    waveform π_half_sig = gaussian(..., π_half_time, ...)

    // Waveforms ultimately supplied to the AODs. We mix our general Gaussian pulse with a sine wave to
    // put a sideband on the signal construction to target the qubit position while maintainig the
    // desired Rabi rate.

>     waveform q1_π_half_sig = mix(π_half_sig, sine(q1_π_half_amp, q1_pos_freq-qubit_freq, 0.0, π_half_time));

Reviewer 3: The order of arguments to the `sine` function doesn't match the definition from the "Waveforms" section.

Reviewer 2: Fixed in 76bccbf.

    waveform q2_π_half_sig = mix(π_half_sig, sine(q2_π_half_amp, q2_pos_freq-qubit_freq, 0.0, π_half_time));
    waveform q3_π_half_sig = mix(π_half_sig, sine(q3_π_half_amp, q3_pos_freq-qubit_freq, 0.0, π_half_time));

>     for τ in [0: T]:

Reviewer: `T` not defined--maybe add to `cal` block?

Reviewer 3: Also the `qubit_freq, Δ, tppi_1, tppi_2,` and `tppi_3`.

Reviewer 2: I left a warning comment in the example about this in ea42e73.

        // Simultaneous π/2 pulses
        play(eom_a_channel, constant(raman_a_amp, π_half_time) , raman_a_frame);
        play(eom_b_channel, constant(raman_b_amp, π_half_time) , raman_b_frame);
        play(aod_channel, q1_π_half_sig, q1_frame);
        play(aod_channel, q1_π_half_sig, q2_frame);
        play(aod_channel, q1_π_half_sig, q3_frame);

        // Time delay all

>         delay(τ/2)

Reviewer: ```suggestion
        delay(τ/2);
```
Does  no args delay all? I know this is true for `barrier`...

Author: Yes, although I think this needs further discussion (as was pointed out in another issue)

            play(eom_a_channel, constant(raman_a_amp, π_half_time) , raman_a_frame);
            play(eom_b_channel, constant(raman_b_amp, π_half_time) , raman_b_frame);
            play(aod_channel, q1_π_half_sig, q1_frame);

        // Barrier all then time delay all
        barrier();
        delay(τ/2);

        // Phase shift the signals by a different amount
        q1_frame.phase += tppi_1 * τ;
        q2_frame.phase += tppi_2 * τ;
        q3_frame.phase += tppi_3 * τ;

        // Simultaneous π/2 pulses
        play(eom_a_channel, constant(raman_a_amp, π_half_time) , raman_a_frame);
        play(eom_b_channel, constant(raman_b_amp, π_half_time) , raman_b_frame);
        play(aod_channel, q1_π_half_sig, q1_frame);
        play(aod_channel, q1_π_half_sig, q2_frame);
        play(aod_channel, q1_π_half_sig, q3_frame);
  }

Multiplexed readout and capture
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this example, we want to perform readout and capture of a pair of qubits, but mediated by a
single physical tx and rx channel. The example is for just two qubits, but works the same for
many (just adding more frames, waveforms, plays, and captures).

.. code-block:: javascript

>   defcal multiplexed_readout_and_capture $0 $1 {

Reviewer: Return type? Seems like it needs to return `bit[2]`

>   defcal multiplexed_readout_and_capture $0 $1 {

Reviewer: ```suggestion
  defcal multiplexed_readout_and_capture $0, $1 {
```

      // the tx/rx channel is the same for $0 and $1

>       channel ro_tx = txch($0, "readout");

Reviewer: ```suggestion
      txchannel ro_tx = txch($0, "readout");
```

>       channel ro_rx = rxch($0, "readout");

Reviewer: ```suggestion
      rxchannel ro_rx = rxch($0, "readout");
```

      // readout frames of different frequencies
      frame q0_frame = newframe(q0_ro_freq, 0); // time 0
      frame q1_frame = newframe(q1_ro_freq, 0); // time 0

      // flat-top readout waveforms
      waveform q0_ro_wf = constant(amp=0.1, l=...);
      waveform q1_ro_wf = constant(amp=0.2, l=...);

      // multiplexed readout
      play(ro_tx, q0_ro_wf, q0_frame);
      play(ro_tx, q1_ro_wf, q1_frame);

      // simple boxcar kernel
      waveform ro_kernel = constant(amp=1, l=...);

      // multiplexed capture

>       // kernel capture(channel chan, waveform ro_kernel, frame capture_frame) -> bit;

Reviewer 3: ```suggestion
      // extern capture(channel chan, waveform ro_kernel, frame capture_frame) -> bit;
```

      bit q0_bit = capture(ro_rx, ro_kernel, q0_frame);
      bit q1_bit = capture(ro_rx, ro_kernel, q1_frame);
      ...
  }

Sample rate collisions
-----------------------

Incommensurate Rates
~~~~~~~~~~~~~~~~~~~~

Since the frame can be played on multiple channels, there may be an issue with sample rates.
For example,

.. code-block:: javascript

  defcal incommensurate_rates_interval $q

>     channel tx0 = txch(0, "tx0"); # sample per 1 ns

Reviewer: Make `txchannel`'s.

Reviewer 3: Offline agreement was to go with `channel`.

    channel tx1 = txch(1, "tx1"); # sample per 2 ns

    waveform wf = gaussian_square(0.1, 13ns, ...);

    play(tx0, wf, driveframe);
    // now driveframe.time is at 13ns
    play(tx1, wf, driveframe); // does not support 13 ns -- either 12ns or 14 ns
  }

The implementation of this behavior is up to the vendor.

Incommensurate Lengths
~~~~~~~~~~~~~~~~~~~~~~

If the samples are defined dt, then playing the same waveform on two different channels
produces

.. code-block:: javascript

  defcal incommensurate_lengths $q

>     channel tx0 = txch(0, "tx0"); # sample per 1 ns

Reviewer: Make `txchannel`'s.

OpenPulse Grammar
=================

OpenQASM allows users to provide the target system's implementation of quantum operations
with ``defcal`` blocks . Calibration grammars are open to extension for system implementors. In
this document, we outline one such grammar, OpenPulse, which maybe selected within a supporting
compiler through the declaration defcalgrammar "openpulse" <VERSION>.

This grammar is motivated by the original OpenPulse specification, a JSON wire-format for
pulse-level quantum programs defined in the paper `Qiskit Backend Specifications for OpenQASM

> and OpenPulse Experiment`_:cite:`mckay2018`. The textual format described

Reviewer 4: Prior to final acceptance, we should check for other work that ought to be cited.

Reviewer: This might be worth adding? https://arxiv.org/pdf/2004.06755.pdf

Reviewer 2: Addressed in upcoming commit 3877695

single NCO).

Frame Construction
~~~~~~~~~~~~~~~~~~

Frames can be constructed using the ``Frame`` command e.g.

.. code-block:: javascript

  frame driveframe = Frame(5e9, 0.0); // Frame(float[size] frequency, angle[size] phase)

would construct a frame with frequency `f`  5 GHz and phase :math:`\theta` 0.0. When
instantiated, the frame time starts at 0. ``Frame``s can also be copied using the
``copy`` command

.. code-block:: javascript

  frame driveframe1 = Frame(5e9, 0.0);
  frame driveframe2 = copy(driveframe1);
  driveframe2.phase = driveframe1.phase + pi/2;

This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` with the same
time as `driveframe1`.

Unless a ``calconfig`` block is used to instantiate frames in global scope, a ``frame`` is
locally scoped . To allow for parameterized access to globally defined frames, we can use
the ``setframe`` and ``getframe`` functions

.. code-block:: javascript

   setframe(str name, frame fr) $q;
   getframe(str name) $q-> frame;

For example, we can parmaeterize frames on different qubits using the wildcard reference
.. code-block:: javascript

>    calconfig {

Reviewer: This offers a solution on scoping. This is a point for discussion...

Reviewer: Review `getframe/setframe` vs `getattr/setattr`, consider qubit wildcards, etc...

Reviewer 3: On a related note, are we also considering the ability to allow for other calibration parameters to skip across `defcal`s?  For example, should I be able to access a calibrated Rabi pulse or qubit frequency?

Reviewer: I think this is a good idea. 

More generally, if we want feature parity with the current OpenPulse, we will need to allow other classical parameters as input to `defcal`'s. An example is doing a Rabi or spectroscopy experiment (input amp or freq). I am inclined to think of these as pulse subroutines rather than gate calibrations. They could then be called directly without reference to a gate. Maybe we should allow this with a different keyword?

Reviewer: After discussion w/ the group, @stevenheidel and @taalexander are going to word towards adding a scoping construct analogous to `box` which may contain inline pulse statements. This allows us to distinguish between two use cases: calibrating a unitary gate and working at a low level of pulse control, at which time the IR is closer to the hardware.

Reviewer: This has been added as the `cal` block implementation. Closing this thread. Thx @stevenheidel, @taalexander!

.. code-block:: javascript

   // Swap phases between two frames
   const temp = frame1.phase;
   frame1.phase = frame2.phase;
   frame2.phase = temp;

Waveforms
---------

Waveforms are of type ``waveform`` and can either be:

- An array of complex samples which define the points for the waveform envelope
- An abstract mathematical function representing a waveform. This will later be
  materialized into a list of complex samples, either by the compiler or the hardware
  using the parameters provided to the pulse template.

A value of type ``waveform`` is retrieved by explicitly constructing the complex samples
or by calling one of the built-in waveform template functions. The latter are initialized by
assigning a ``waveform`` to the result of a (kernel) function call. Note that each of these
kernel functions takes a type ``length`` as a first argument, since waveforms need to have a
definite length. Using the hardware dependent ``dt`` unit is recommended for this length,
since the compiler may need to down-sample a higher precision waveform to physically realize it.

.. code-block:: javascript

   // arbitrary complex samples

>    waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];

Reviewer: Note: Will need to define array and complex types either in OpenQASM or specifically at the OpenPulse level. There have been discussion for both at the OpenQASM level

Reviewer 4: I would support adding both at the OpenQASM level.

   kernel sech(complex[size] amp, length l, length sigma) -> waveform;

   // amp is waveform amplitude at center
   // l is the overall length of the waveform
   // square_width is the width of the square waveform component
   // sigma is the standard deviation of waveform
   kernel gaussian_square(complex[size] amp, length l, length square_width, length sigma) -> waveform;

   // amp is waveform amplitude at center
   // l is the overall length of the waveform
   // sigma is the standard deviation of waveform
   // beta is the Y correction amplitude, see the DRAG paper
   kernel drag(complex[size] amp, length l, length sigma, float[size] beta) -> waveform;

   // amp is waveform amplitude
   // l is the overall length of the waveform
   kernel constant(complex[size] amp, length l) -> waveform;

   // amp is waveform amplitude
   // l is the overall length of the waveform
   // frequency is the frequency of the waveform
   // phase is the phase of the waveform
   kernel sine(complex[size] amp, length  l, float[size] frequency, angle[size] phase) -> waveform;

We can manipulate the ``waveform`` types using the following signal processing functions to produce
new waveforms

> 

Reviewer: Discussion: Is it worth keeping all of the processing functions or postponing some for future versions?

Reviewer 3: I think some of these could be nixed. The ones I don't see an immediate use for are
- `piecewise`
- `offset`
- `conj` (can be achieved w/ `phase_shift`)
- `re`
- `im`
- `abs`

Reviewer 5: `piecewise` might have use for "flat-topped" pulses where we put together a rise, flat-top, and fall.

Reviewer: Addressed in https://github.com/Qiskit/openqasm/pull/184/commits/2332cf7fa7af9e266092a6bf536048e6b867553a.

  This is equivalent to the product signal :math:`wf(t_i) = wf_1(t_i) \times wf_2(t_i)`
- ``sum(waveform wf1, waveform wf2)`` -> waveform - Sum two input waveforms to produce a new waveform.
- ``piecewise(waveform wf0, waveform wf1)`` -> waveform - Output waveform.
- ``offset(waveform wf, complex amount)`` -> waveform - Offset the input waveform by an amount.
- ``scale(waveform wf, complex factor)`` -> waveform - Scale the input waveform by a factor.
- ``conj(waveform wf) -> waveform`` - Conjugate the input waveform.
- ``re(waveform wf) -> waveform`` - Real component of input waveform.
- ``im(waveform wf) -> waveform`` - Imaginary component of input waveform.
- ``abs(waveform wf) -> waveform`` - Transform waveform as norm of input. waveform
- ``phase_shift(waveform wf, angle ang) -> waveform`` - Signal with relative phase, ang.

Play instruction
----------------

Waveforms are scheduled using the ``play`` instruction. These instructions may
only appear inside a ``defcal`` block and have three required parameters:

- the channel on which to play the pulse
- a value of type ``waveform`` representing the waveform envelope
- the frame to use for the pulse

.. code-block:: javascript

>   play(channel chan, waveform wfm, frame fr)

Reviewer: Potential modification: Define a virtual channel type `vxchannel` (or other name). Tie channels to frames in the form of virtual channels, eg., `vxchannel chan = virtual(channel, frame)` and do `play(vxchannel vc_chan, waveform wfm)`. Virtual channel tracks time based on the frame time.

Reviewer 3: That would be probably be the most often used case, but I wonder if the minor convenience of not having to write both the channel and frame in a `play` instruction worth the complexity (however small) of adding yet another type?

Reviewer: That is a fair point. One benefit of the virtual channel approach is that timing is tracked in a linear way and never needs to be done across channels...

Reviewer: We could also tie the virtual channel to qubits rather than doing it in `setframe/getframe`.

Reviewer: This issue was discussed with the group. It was decided that we leave things as they are for the following reasons:
- Issue of updating frames: when we update the channel, does it update the frame or do we have to do this separately?
- Frame should contain timing truth
- Geometric gates are less clear w/ virtual channels

small differences.

The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
instruction may also take a list of frames instead of a list of qubits and aligns the time
of the clocks given as arguments.

``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
meaning that those clocks are guaranteed to be aligned at the start of the block.
These blocks also need to have a well-defined length, similar to the ``boxas`` block.

.. code-block:: javascript

   waveform p = ...; // some 100dt waveform

   calconfig {
     setframe("frame",  Frame(5.0e9, 0)) $0;
     setframe("frame",  Frame(6.0e9, 0)) $1;
   }

   defcal aligned_gates {
     // driveframe1 and driveframe2 used in this defcal, so clocks are aligned
     frame driveframe1 = getframe("frame") $0;
     frame driveframe2 = getframe("frame") $1;
     play(tx0, p, driveframe1);
     delay[20dt] driveframe1;
     // Clocks now unaligned by 120dt, so we use a `barrier` to re-align
     barrier(driveframe1, driveframe2);
     // `driveframe2` will now play a pulse 120dt after `driveframe1` finishes playing
     play(tx0, p, driveframe2);
   }

> Examples

Reviewer: Is a spectroscopy example worth adding?

Reviewer 5: With the new `cal` block would it look something like this?

```
cal {
    frame qubit_drive = newframe(5.0e9, 0);
}

defcal saturation_pulse $0 {
    txchannel d0 = txch($0, "drive")
    play(d0, constant(0.1, 100e-6))
}

// sweep 100 MHz around the qubit drive frequency

cal {
    qubit_drive.frequency -= 50e6
}

for i in [1:100] {
    cal {
        qubit_drive.frequency += 1e6
    }

    saturation_pulse $0
    measure $0
    
}
```

        play(aod_channel, q1_π_half_sig, q3_frame);
  }

Multiplexed readout and capture
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this example, we want to perform readout and capture of a pair of qubits, but mediated by a
single physical tx and rx channel. The example is for just two qubits, but works the same for
many (just adding more frames, waveforms, plays, and captures).

.. code-block:: javascript

  defcal multiplexed_readout_and_capture $0 $1 {

      // the tx/rx channel is the same for $0 and $1
      tx_channel ro_tx = txch($0, "readout");
      rx_channel ro_rx = rxch($0, "readout");

      // readout frames of different frequencies
      q0_frame = Frame(q0_ro_freq, 0); // time 0
      q1_frame = Frame(q1_ro_freq, 0); // time 0

      // flat-top readout waveforms
      waveform q0_ro_wf = constant(amp=0.1, l=...);
      waveform q1_ro_wf = constant(amp=0.2, l=...);

      // multiplexed readout
      play(ro_tx, q0_ro_wf, q0_frame);
      play(ro_tx, q1_ro_wf, q1_frame);

      // simple boxcar kernel
      waveform ro_kernel = constant(amp=1, l=...);

      // multiplexed capture
      // kernel capture(channel chan, waveform ro_kernel, frame capture_frame) -> bit;

>       bit q0_bit = capture(ro_rx, ro_kernel, q0_frame);

Reviewer 5: I don't think we can reuse the frame for `capture` because the `capture` would be scheduled after the `play`.

Reviewer 3: Yes, I think we need to split this out into stimulus and capture frames. 
```
// readout stimulus and capture frames of different frequencies
frame q0_stimulus_frame = newframe(q0_ro_freq, 0); // time 0
frame q0_capture_frame = newframe(q0_ro_freq, 0); // time 0
frame q1_stimulus_frame = newframe(q1_ro_freq, 0); // time 0
frame q1_capture_frame = newframe(q1_ro_freq, 0); // time 0
```

Then we would want to use a `barrier` before capturing on the `q#_capture_frame`s

The frame is composed of three parts:

1. A frequency ``frequency`` of type ``float``.
2. A phase ``phase`` of type ``angle``.
3. A time which is manipulated implicitly and cannot be modified other
   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
   is determined by the channel on which the frame is played (see :ref:`Timing` section).

The ``frame`` type is a virtual resource and the exact precision of these parameters is
hardware specific. It is thus up to the compiler to choose how to implement the required
transformations to physical resources in hardware (e.g. mapping multiple frames to a
single NCO).

Frame Construction
~~~~~~~~~~~~~~~~~~

Frames can be constructed using the ``newframe`` command e.g.

.. code-block:: javascript

  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)

would construct a frame with a frequency of 5 GHz and a phase of 0.0. When
instantiated, the frame time starts at 0. ``frame``\s can also be copied using the
``copyframe`` command

.. code-block:: javascript

  frame driveframe1 = newframe(5e9, 0.0);
  frame driveframe2 = copyframe(driveframe1);

>   driveframe2.phase = driveframe1.phase + pi/2;

Reviewer 4: Not unique here (we have this elsewhere in OpenQASM, too, but this example involves some kind of implicit casting of `pi/2` to an `angle` type to make this meaningful.

Author: Would it be reasonable to always consider `pi` to be an angle?

Reviewer 4: No, but we can address this in the implicit casting rules defined by the types WG.

phase. Again, the exact precision of these calculations is hardware specific.

Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
``driveframe``:

.. code-block:: javascript

   // Example 1: Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
   cal {
     driveframe.phase += pi/4;
   }

   // The following is an example only. Frames as arrays has not been agreed on.
   // This conceptually must be compile-time arrays and treat qubits as indices
   // which also has not been well-defined. We are exploring other solutions to
   // the problem of mapping qubits to pulse-level resources.

   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
   cal {
     array[frame, 8] rz_frames;
     frame[0] = newframe(...);
     // and so on
   }

   defcal rz(angle[20] theta) $q {
     rz_frames[q].phase -= theta;
   }

Manipulating frames based on the state of other frames is also permitted:

.. code-block:: javascript

   // Swap phases between two frames
   const temp = frame1.phase;
   frame1.phase = frame2.phase;
   frame2.phase = temp;

Waveforms
---------

> Waveforms are of type ``waveform`` and can either be:

Reviewer 4: Why do we need a `waveform` type instead of using array types like `array[complex[float], len]`?

Author: Discussed in meeting, but for the github record two reasons:
1. It is useful for the compiler and control hardware to work with "gaussian" and a few parameters rather than a list of values that happen to be a gaussian
2. Array lengths in OpenQASM are static, so there'd be no way to pass in length as an input

        play(m0, meas_wf, stimulus_frame);
        // Align measure and capture channels
        barrier(stimulus_frame, capture_frame);
        // Capture transmitted data after interaction with measurement resonator
        // extern capture(channel chan, length duration, frame capture_frame) -> waveform;
        waveform raw_output = capture(cap0, 16000dt, capture_frame);

        // Kernel and discriminate
        complex[32] iq = boxcar(raw_output);
        bit result = discriminate(iq);

        return result;
    }

Timing
------

Each frame maintains its own "clock". When a pulse is played the clock for
that frame advances by the length of the pulse.

For frames, everything behaves analogous to qubits in the
`Delays <delays.html>`_ section of this specification. There are however some
small differences.

The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
instruction may also take a list of frames instead of a list of qubits and aligns the time
of the clocks given as arguments.

``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
meaning that those clocks are guaranteed to be aligned at the start of the block.

> These blocks also need to have a well-defined length, similar to the ``boxas`` block.

Reviewer 2: Remove boxas - 35aad58.

.. code-block:: javascript

   cal {
     waveform p = ...; // some 100dt waveform
     frame driveframe1 = newframe(5.0e9, 0);
     frame driveframe2 = newframe(6.0e9, 0);
   }

>    defcal aligned_gates {

Reviewer 4: Is this supposed to be just a `cal` block? I don't know what a `defcal` is without qubit arguments.

Reviewer 2: Added qubits in c89f99a.

other `efforts in the field`_:cite:`alexanderQiskitPulseProgramming2020a,DiCarloLabDelftPycQEDPy32021,ExperimentalMicroarchitectureSuperconducting,nguyenEnablingPulselevelProgramming2020,QuillangQuil2021`.

The textual format described here has several advantages over the original JSON format:

- Improved readability.
- Pulse timing is based on instruction ordering and works with programs containing branching
  control flow.
- Reusable gate calibrations enable more succinct calibration descriptions.
- Pulse definitions are declared as a calibration for individual circuit instructions attached to
  physical qubits enabling the microcoding of gate level operations.
- Richer ability to compose complex pulses through natural DSP-like operations.
- Clearly defined relationship between pulses, channels and the phases of compilation to
  hardware resources.
- Use of multiple frames on a single channel at the same time (see :ref:`Geometric gate`).

Openpulse provides a flexible programming model that should extend to many quantum control schemes
and hardware. At the core of the OpenPulse grammar are the concepts of ``frames``, ``waveforms`` and ``channels``.

> ``frames`` are used to schedule the playing of a ``waveform`` or capturing of data

Reviewer 2: Updated in e94fabe. Still not really happy with it, nor the dual purpose of frames for tracking time and frequency/phase. Seems hard to conceptually explain. Would appreciate suggestions on how to fix.

via a ``channel`` on the target device. A ``channel`` is a software abstraction which allows the
programmer to be agnostic to complexities of the device's underlying pulse generation hardware. It
is the responsibility of the target device's compiler to map ``frames`` to the applied ``channel``s in
target hardware.

Channels
--------

> Channels map to a hardware resource, which can play pulses to manipulate a qubit

Reviewer 2: Made an attempt on this in c957ba4.

> or capture data from the qubit to perform a measurement. There is a many-to-many

Reviewer 2: Yes, I agree. Updated language in b53776f.

relationship between qubits and channels. One qubit may have multiple channels
connecting to it. Pulses on different channels would have different physical

> interactions with that qubit. A channel may also have many qubits. For instance,

Reviewer 2: Addressed in f8ccf34.

of string name and qubits.

.. code-block:: c

    getchannel(str name, qubit q0, ..., qubit qn) -> channel  // get a channel

The qubits must be **physical** qubits. Furthermore, ordering of qubits is important. For instance,
``getchannel("control", $0, $1)`` and ``getchannel("control", $1, $0)`` may be used to implement distinct
cross-resonance gates. It is also possible to access a channel by its full name, without supplying
any qubits, if that has been implemented by the vendor. For instance, ``getchannel("<channel_name>")``
may refer to a transmit channel with an arbitrary name.

.. code-block:: c

    channel d0 = getchannel("drive", $0);  // channel for driving at qubit $0's freq
    channel cr1_2 = getchannel("coupler", $1, $2);  // channel for a coupler between two qubits
    channel m2 = getchannel("measure", $2);  // channel for transmitting measurement stimulus

    channel global_field = getchannel("global_field"); // channel not associated with any qubits

    // capture channels for capturing qubits $0 and $1
    channel cap0 = getchannel("capture", $0);
    channel cap1 = getchannel("capture", $1);

Frames
------

> When interacting with qubits, it is powerful to track of a frame of reference, akin to the rotating

Reviewer 2: Fixed in d4cd932.

  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
  defined in the rotating frame with the effect being an equivalent application in the lab
  frame (i.e. with the carrier supplied by the ``frame``). Another motivation is to more naturally
  implement a "virtual Z-gate", which does not require a physical pulse but rather shifts the phase
  of all future pulses on that frame.

The frame is composed of three parts:

1. A frequency ``frequency`` of type ``float``.
2. A phase ``phase`` of type ``angle``.
3. A time which is manipulated implicitly and cannot be modified other
   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
   is determined by the channel on which the frame is played (see :ref:`Timing` section).

The ``frame`` type is a virtual resource and the exact precision of these parameters is
hardware specific. It is thus up to the compiler to choose how to implement the required
transformations to physical resources in hardware (e.g. mapping multiple frames to a

> single NCO).

Reviewer 2: Numerically controlled oscillator. Updated with name and link in bc82d56.

Frame Construction
~~~~~~~~~~~~~~~~~~

Frames can be constructed using the ``newframe`` command e.g.

.. code-block:: javascript

  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)

would construct a frame with a frequency of 5 GHz and a phase of 0.0. When
instantiated, the frame time starts at 0. ``frame``\s can also be copied using the
``copyframe`` command

.. code-block:: javascript

  frame driveframe1 = newframe(5e9, 0.0);
  frame driveframe2 = copyframe(driveframe1);
  driveframe2.phase = driveframe1.phase + pi/2;

This will generate a :math:`\pi/2` phase incremented copy of ``driveframe1`` with the same

> Frame manipulation

Reviewer 2: Above the frame was incremented.

   cal {
     driveframe.phase += pi/4;
   }

   // The following is an example only. Frames as arrays has not been agreed on.
   // This conceptually must be compile-time arrays and treat qubits as indices
   // which also has not been well-defined. We are exploring other solutions to
   // the problem of mapping qubits to pulse-level resources.

   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
   cal {
     array[frame, 8] rz_frames;
     frame[0] = newframe(...);
     // and so on
   }

   defcal rz(angle[20] theta) $q {
     rz_frames[q].phase -= theta;
   }

Manipulating frames based on the state of other frames is also permitted:

.. code-block:: javascript

   // Swap phases between two frames
   const temp = frame1.phase;
   frame1.phase = frame2.phase;
   frame2.phase = temp;

Waveforms
---------

Waveforms are of type ``waveform`` and can either be:

- An array of complex samples which define the points for the waveform envelope
- An abstract mathematical function representing a waveform. This will later be
  materialized into a list of complex samples, either by the compiler or the hardware

>   using the parameters provided to the ``extern`` defined pulse template.

Reviewer 6: Maybe add a sentence based on offline discussion: like other `extern` functions, these will be compiled. But for static waveforms, the optimizing compiler should decide to execute this at compile time and load the waveform into memory once. For dynamic waveforms, the compiler just compiles and links this, to be executed at runtime.

Author: Great suggestion, added.

measurement. Exploring the response to sweeps of pulse frequency, time, amplitude, or even
multi-dimensional sweeps reveals spectroscopic information about the qubit transition frequencies
and the drive strength. We describe these circuits with a mixture of conventional OpenQASM for the
simple pulse and measure sequence and step into `cal` blocks to access pulse level control. We
assume that the OpenQASM text is generated by some higher level language bindings and we only write
into the program the sweep where we can take advantage of the execution speed of sweeping as part of
the program.

**Qubit Spectroscopy**

Here we want to sweep the frequency of a long pulse that saturates the qubit transition.

.. code-block:: javascript

  // sweep parameters would be programmed in by some higher level bindings
  const float frequency_start = 4.5e9;
  const float frequency_step = 1e6
  const int frequency_num_steps = 301;

  // define a long saturation pulse of a set length and amplitude
  defcal saturation_pulse $0 {
      // assume channel and frame can be linked from a vendor supplied `cal` block

>       play(tx0, constant(0.1, 100e-6), driveframe);

Reviewer 7: `driveframe` is not defined.

Reviewer 3: Comment says 
>  // assume channel and frame can be linked from a vendor supplied `cal` block

The textual format described here has several advantages over the original JSON format:

- Improved readability.
- Pulse timing is based on instruction ordering and works with programs containing branching
  control flow.
- Reusable gate calibrations enable more succinct calibration descriptions.
- Pulse definitions are declared as a calibration for individual circuit instructions attached to
  physical qubits enabling the microcoding of gate level operations.
- Richer ability to compose complex pulses through natural DSP-like operations.
- Clearly defined relationship between pulses, channels and the phases of compilation to
  hardware resources.
- Use of multiple frames on a single channel at the same time (see :ref:`Geometric gate`).

Openpulse provides a flexible programming model that should extend to many quantum control schemes
and hardware. At the core of the OpenPulse grammar are the concepts of ``channel``s ``waveform``s and, ``frame``s.
A ``channel`` is a software abstraction which allows the programmer to be agnostic to the complexities
of the device's underlying pulse generation or capture hardware, representing an input or output channel to various components

> controlling qubits such as microwave lines. A ``waveform`` is a time-dependent envelope that is emitted a channel in

Reviewer 4: ```suggestion
controlling qubits such as microwave lines. A ``waveform`` is a time-dependent envelope that is emitted on a channel in
```

phase. Again, the exact precision of these calculations is hardware specific. The frequency
and phase updates will applied immediately at the current time of the frame.

Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
``driveframe``:

.. code-block:: javascript

  ...

   // Example 1: Shift phase of the "drive" frame by pi/4, to realize a virtual rz gate with angle -pi/4
   cal {
     driveframe.phase += pi/4;
   }

   // The following is an example only. Frames as arrays has not been agreed on.
   // This conceptually must be compile-time arrays and treat qubits as indices
   // which also has not been well-defined. We are exploring other solutions to
   // the problem of mapping qubits to pulse-level resources.

   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
   cal {
     array[frame, 8] rz_frames;
     frame[0] = newframe(...);
     // and so on
   }

   defcal rz(angle[20] theta) $q {
     rz_frames[q].phase -= theta;
   }

Manipulating frames based on the state of other frames is also permitted:

.. code-block:: javascript

   // Swap phases between two frames

>    const temp = frame1.phase;

Reviewer 4: ```suggestion
   const angle temp = frame1.phase;
```
?

Using the hardware dependent ``dt`` unit is recommended for this duration,
since otherwise the compiler may need to down-sample a higher precision
waveform to physically realize it.

Like other extern functions, ``extern waveform`` functions will be compiled.
But for static waveforms, the optimizing compiler should decide to execute this
at compile time and load the waveform into memory once.
For dynamic waveforms, the compiler just compiles and links this, to be executed at runtime.
We provide the ``waveform`` type in addition to the complex list of samples to
provide more context to compilers and hardware. For example, some hardware pulse
generators may have optimized implementations of common pulse shapes like gaussians.
Providing structured gaussian parameters instead of the materialized list of complex
samples provides optimization opportunities that wouldn't be available otherwise.

.. code-block:: javascript

   // arbitrary complex samples
   waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];

   // amp is waveform amplitude at center
   // d is the overall duration of the waveform
   // sigma is the standard deviation of waveform

>    extern gaussian(complex[size] amp, duration d, duration sigma) -> waveform;

Reviewer 4: Here, and basically in all the example below, these `complex` types need to be `complex[float[size]]`.

new waveforms (this list may be updated as more functionality is required).

.. code-block:: javascript

    // Multiply two input waveforms entry by entry to produce a new waveform
    // :math:`wf(t_i) = wf_1(t_i) \times wf_2(t_i)`
    mix(waveform wf1, waveform wf2) -> waveform;

    // Sum two input waveforms entry by entry to produce a new waveform
    // :math:`wf(t_i) = wf_1(t_i) + wf_2(t_i)`
    sum(waveform wf1, waveform wf2) -> waveform;

    // Add a relative phase to a waveform (ie multiply by :math:`e^{\imag \theta}`)
    phase_shift(waveform wf, angle ang) -> waveform;

    // Scale the amplitude of a waveform's samples producing a new waveform
    scale(waveform wf, float factor) -> waveform;

Play instruction
----------------

Waveforms are scheduled using the ``play`` instruction. These instructions may
only appear inside a ``defcal`` block and have three required parameters:

- The channel on which to play the pulse.
- A value of type ``waveform`` representing the waveform envelope.
- The frame to use for the pulse.

.. code-block:: javascript

  play(channel chan, waveform wfm, frame fr)

For example,

.. code-block:: javascript

>   defcal play_my_pulses {

Reviewer 4: Do we allow `defcals` without qubit arguments?

.. code-block:: javascript

   // Minimum requirement
   extern capture(channel chan, frame output);

   // A capture command that returns an iq value
   extern capture(channel chan, waveform filter, frame output) -> complex[32];

   // A capture command that returns a discrimnated bit
   extern capture(channel chan, waveform filter, frame output) -> bit;

   // A capture command that returns a raw waveform data
   extern capture(channel chan, duration len, frame output) -> waveform;

   // A capture that returns a count e.g. number of photons detected
   kernel capture(channel chan, duration len, frame output) -> int

The return type of a ``capture`` command varies. It could be a raw trace, ie., a
list of samples taken over a short period of time. It could be some averaged IQ
value. It could be a classified bit. Or it could even have no return value,
pushing the results into some buffer which is then accessed outside the program.

For example, the ``capture`` instruction could return raw waveform data that is then
discriminated using user-defined boxcar and discrimination ``extern``\s.

.. code-block:: javascript

    // Use a boxcar function to generate IQ data from raw waveform

>     extern boxcar(waveform input) -> complex[64];

Reviewer 4: ```suggestion
    extern boxcar(waveform input) -> complex[float[64]];
```

    // Use a linear discriminator to generate bits from IQ data

>     extern discriminate(complex[64] iq) -> bit;

Reviewer 4: ```suggestion
    extern discriminate(complex[float[64]] iq) -> bit;
```

    defcal measure $0 -> bit {
        // Define the channels
        channel m0 = getchannel("measure", $0);
        channel cap0 = getchannel("capture", $0);

        // Force time of carrier to 0 for consistent phase for discrimination.
        frame stimulus_frame = newframe(5e9, 0);
        frame capture_frame = newframe(5e9, 0);

        // Apply measurement stimulus
        waveform meas_wf = gaussian_square(1.0, 16000dt, 262dt, 13952dt);

        // Play the stimulus
        play(m0, meas_wf, stimulus_frame);
        // Align measure and capture channels
        barrier stimulus_frame, capture_frame;
        // Capture transmitted data after interaction with measurement resonator
        // extern capture(channel chan, duration duration, frame capture_frame) -> waveform;
        waveform raw_output = capture(cap0, 16000dt, capture_frame);

        // Kernel and discriminate

>         complex[32] iq = boxcar(raw_output);

Reviewer 4: ```suggestion
        complex[float[32]] iq = boxcar(raw_output);
```

  cal {
    channel tx0 = getchannel("drive", $0); // sample per 1 ns
    channel tx1 = getchannel("drive", $1); // sample per 2 ns
    waveform p = ...; // some 100ns waveform
    frame driveframe1 = newframe(5.0e9, 0);
    frame driveframe2 = newframe(6.0e9, 0);
  }

  defcal two_qubit_gate $1 $2 {
    play(tx0, wf, driveframe1);
    play(tx1, wf, driveframe2);
  }

  defcal single_qubit_gate $1 {
    play(tx0, wf, driveframe1);
  }

  single_qubit_gate $1;
  // Implicit alignment of `driveframe1` and `driveframe2` when entering `two_qubit_gate` block
  two_qubit_gate $1 $2;

Incommensurate Times
~~~~~~~~~~~~~~~~~~~~

A frame's "clock" may be at a time incomensurate with the sample rate of the
underlying ``channel`` in the next ``play`` or ``capture`` instruction. For example,

.. code-block:: javascript

  defcal incommensurate_rates_interval $q
    channel tx0 = getchannel("tx0", 0); # sample per 2 ns
    waveform wf = gaussian_square(0.1, 12ns, ...);
    driveframe = newframe(4.4e9, 0.0);

    delay[13ns] driveframe;

    // driveframe.time is at 13ns, but the tx0 channel can only realize
    // either 12ns or 14 ns time point.
    play(tx0, wf, driveframe);
  }

> The expected behavior (e.g. to implictly cast to the next realizable time

Author: ```suggestion
The expected behavior (e.g. to implicitly cast to the next realizable time
```

OpenPulse Grammar
=================

> *The OpenPulse grammar is still in active development and is liable to change.*

Reviewer 8: ```suggestion
*The OpenPulse grammar is still in active development and is liable to change. If you are working on an implementation and find this specification unclear or not supporting your use-cases, please join our community effort to improve pulse-level support in OpenQasm.*
```
